apis/v1alpha1/validation.go

```diff
@@ -33,11 +35,64 @@ func ValidateResourceMapping(m *ResourceMapping) (vErr error) {
 			vErr = errors.Join(vErr, fmt.Errorf("invalid owner: %w", err))
 		}
 	}
-	if m.Resource.Provider == "" {
-		vErr = errors.Join(vErr, fmt.Errorf("empty resource provider"))
-	}
+
 	if _, ok := m.Annotations.AsMap()[AnnotationKeyAssetInfo]; ok {
 		vErr = errors.Join(vErr, fmt.Errorf("reserved key is included: %s", AnnotationKeyAssetInfo))
 	}
+
+	if err := validateResource(m.Resource); err != nil {
+		vErr = errors.Join(vErr, fmt.Errorf("invalid Resourc: %w", err))
+	}
+
 	return
 }
+
+func validateResource(r *Resource) (vErr error) {
+	if r.Name == "" {
+		vErr = errors.Join(vErr, fmt.Errorf("empty resource name"))
+	}
+
+	if r.Provider == "" {
+		vErr = errors.Join(vErr, fmt.Errorf("empty resource provider"))
+	}
+
+	if err := validateAndNormalizeSubscope(r); err != nil {
+		vErr = errors.Join(vErr, fmt.Errorf("invalid subscope: %w", err))
+	}
+
+	return
+}
+
+func validateAndNormalizeSubscope(r *Resource) error {
+	if r.Subscope == "" {
+		return nil
+	}
+
+	u, err := url.Parse(r.Subscope)
+	if err != nil {
+		return fmt.Errorf("failed to parse subscope string: %w", err)
+	}
+
+	_, err = url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return fmt.Errorf("failed to parse qualifier string: %w", err)
+	}
+
+	// check if key value pairs are in alphabetical order
+	// and if key is in lower case
+	s := strings.Split(u.RawQuery, "&")
```

Comment:
The naming is really confusion and the validation logic is difficult to understand. I'd suggest you renaming the variable with clear meaning instead of using `s` or `s1`. For example, you can use `hierarchyStr`, qualifierStr`, `qualifierKey`, `qualifierValue`, `hierarchyKey` `hierarchyValue`. In addition, do you need any validation for the `hierarchyKey`?

Reply:
@yolocs  By the way,I doubt whether we need to do such validation here like key value pairs are in alphabetical order? I think making keys all lower case could bring some value. But I didn't get the point that  key value pairs are in alphabetical order

Reply:
making the key value pairs in alphabetical put can help make sure the qualifier string will be the same as long as the key value pairs are the same, so we won't have cases like `key1=value1&key2=value2` and `key2=value2&key1=value1` both exists.

As for validating hierarchyKey, if we decided not to check key is in lower case, and we can't validate the resource hierarchy, I don't think we need any validation.

Reply:
Not sure I see the value of enforcing lower case keys.

Reply:
Discussed offline, I think we only need to enforce the key value pairs are in alphabetical order.